Show variation picker in case of variable subscription

samiuelson opened the pull request:
## Bug fix: Adding variable subscription to an order

Closes: #8730
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
**Before**: In case of tapping on the variable subscription product the variation picker was not shown. 
As a result, it was impossible to add a subscription variation to an order.

**After**: In case of tapping on the variable subscription product the variation picker is shown now. It's possible to add subscription variation to an order.

### Testing instructions
1. Create product of type: `variable subscription`. Ensure it has at least one variation available (otherwise it's not displayed on the product list).
2. Create new order and tap "+ Add products".
3. Verify variable subscription product is displayed and tapping on it redirects to the variation picker.
4. Select a variation, go back and confirm selection.
5. Verify the subscription variation is added to the order.

### Visual changes
<!-- Include before and after images or gifs when appropriate. -->
| Before  | After |
| ------------- | ------------- |
| <video src=https://user-images.githubusercontent.com/4527432/231437083-1982bc81-bb62-4d44-b3a2-cab35c2b72d6.mp4/>  | <video src=https://user-images.githubusercontent.com/4527432/231436602-56c00508-c863-4361-b5db-3aeb25c25e03.mp4/>  |

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Show variation picker in case of variable subscription

samiuelson committed:
Update RELEASE-NOTES.txt

samiuelson committed:
Fix selection state computation

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorViewModel.kt:
```diff
@@ -280,7 +281,7 @@ class ProductSelectorViewModel @Inject constructor(
     fun onProductClick(item: ProductListItem, productSourceForTracking: ProductSourceForTracking) {
         val productSource = updateProductSourceIfSearchIsEnabled(productSourceForTracking)
-        if (item.type == VARIABLE && item.numVariations > 0) {
+        if ((item.type == VARIABLE || item.type == VARIABLE_SUBSCRIPTION) && item.numVariations > 0) {
```
❓ On line no 293. Do we need to add the below check?
`|| item.type != VARIABLE_SUBSCRIPTION`

AnirudhBhat replied:
❓ Similarly on line no 517. Do we need to add the below check?

`|| product.productType == VARIABLE_SUBSCRIPTION`

AnirudhBhat replied:
❓ Similarly on line no 222. I feel we need to add the below check

`|| productType == VARIABLE_SUBSCRIPTION`

The iOS video shown in the [ticket](https://github.com/woocommerce/woocommerce-android/issues/8730) has `In stock . $50.00 . 3 variations` for a variable subscription. Android do not display this information.

samiuelson replied:
Thank you for the review @AnirudhBhat 🙇🏼 

I've fixed the above bugs with 5ff6b84.

AnirudhBhat reviewed (commented):
Thanks for working on this @samiuelson. I've left a few comments (all minor) that I think we should give thought to before merging the PR. Let me know what you think of those.

samiuelson committed:
Merge branch 'trunk' into bug/8730-variable-subscription-picker

samiuelson committed:
Fix variable subscription items UI model

AnirudhBhat commented:
@samiuelson Thank you for the fix! Sorry, I should have mentioned this earlier. Some of the tests need some modification as well.

`given no variable products with no variations restriction, when view model created, should not show variable products with no variations`
`given multiple restrictions, when view model created, should show correct products`

The above tests need to assert on `variable_subscriptions` as well after the recent change. WDYT?

Also, we could write some tests for [this change](https://github.com/woocommerce/woocommerce-android/pull/8798/commits/5ff6b84a49ea4091e4fbf986c928212307057e44#diff-55e3bd3e7770a782c98571d34db440213816ac7d5820cb9a00019f3bbade5e33R220) to verify that the `stockStatus` returns the proper string and price. We can also take this in a separate PR if you think it's worth writing the test.
```
val stockStatus = when (stockStatus) {
            InStock -> {
                if (isVariable()) {
                    resourceProvider.getString(
                        string.product_stock_status_instock_with_variations,
                        numVariations
                    )
                } else {
                    getStockStatusLabel()
                }
            }
            NotAvailable, is Custom -> null
            else -> resourceProvider.getString(stockStatus.stringResource)
        }

```

samiuelson committed:
Update unit tests

samiuelson committed:
Merge branch 'trunk' into bug/8730-variable-subscription-picker

samiuelson committed:
Add test for stock and price labels

samiuelson commented:
> @samiuelson Thank you for the fix! Sorry, I should have mentioned this earlier. Some of the tests need some modification as well.
> 
> `given no variable products with no variations restriction, when view model created, should not show variable products with no variations` `given multiple restrictions, when view model created, should show correct products`
> 
> The above tests need to assert on `variable_subscriptions` as well after the recent change. WDYT?
> 
> Also, we could write some tests for [this change](https://github.com/woocommerce/woocommerce-android/pull/8798/commits/5ff6b84a49ea4091e4fbf986c928212307057e44#diff-55e3bd3e7770a782c98571d34db440213816ac7d5820cb9a00019f3bbade5e33R220) to verify that the `stockStatus` returns the proper string and price. We can also take this in a separate PR if you think it's worth writing the test.
> 
> ```
> val stockStatus = when (stockStatus) {
>             InStock -> {
>                 if (isVariable()) {
>                     resourceProvider.getString(
>                         string.product_stock_status_instock_with_variations,
>                         numVariations
>                     )
>                 } else {
>                     getStockStatusLabel()
>                 }
>             }
>             NotAvailable, is Custom -> null
>             else -> resourceProvider.getString(stockStatus.stringResource)
>         }
> ```

Thanks for the suggestion! I've updated existing tests with 655b221 and added new tests for stock and price labels with 4c504bd.
